hyppo/independence/friedmanRafsky.py

```diff
@@ -0,0 +1,297 @@
+import random
+from numba import jit
+
+import numpy as np
+
+from .base import IndependenceTest, IndependenceTestOutput
+
+
+class FriedmanRafsky(IndependenceTest):
+
+    r"""
+    Friedman-Rafksy (FR) test statistic and p-value.
+    This is a multivariate extension of the Wald-Wolfowitz
+    runs test for randomness. The normal concept of a 'run'
+    is replaced by a minimum spanning tree calculated between
+    the points in respective data sets with edge weights defined
+    as the Euclidean distance between two such points. After MST
+    has been determined, all edges such that both corresponding
+    nodes do not belong to the same class are severed and the
+    number of independent resulting trees is counted. This test is
+    consistent against similar tests
+    :footcite:p:`GSARMultivariateAdaptationofWaldWolfowitzTest`
+    """
+
+    def __init__(self, **kwargs):
+
+        IndependenceTest.__init__(self, **kwargs)
+
+    class Graph:
+
+        r"""
+        Helper class that finds the MST for a given dataset using
+        Kruskal's MST algorithm. Takes in node lables and edge weights
+        before returning all pairs of connected node labels in the
+        resultant MST.
+        Parameters
+        ----------
+        i,j,weight : int, int, float
+            Input node indeces i, j, and corresponding Euclidean distance
+            edge weight between them.
+        Returns
+        -------
+        result : list
+            List of pairs of nodes connected in final MST.
+        """
+
+        def __init__(self2, vertex):
+
+            self2.V = vertex
+            self2.graph = (
+                []
+            )  # Empty matrix for holding vertices and weights connecting them
+
+        def add_edge(self2, v1, v2, w):
+
+            self2.graph.append(
+                [v1, v2, w]
+            )  # Add method for creating edges between vertices
+
+        def search(
+            self2, parent, i
+        ):  # Method for determining location of vertex in existing tree
+
+            if parent[i] == i:
+                return i
+
+            return self2.search(parent, parent[i])
+
+        def apply_union(
+            self2, parent, rank, x, y
+        ):  # Method for deleting and merging branches
+
+            xroot = self2.search(parent, x)
+            yroot = self2.search(parent, y)
+            if rank[xroot] < rank[yroot]:
+                parent[xroot] = yroot
+            elif rank[xroot] > rank[yroot]:
+                parent[yroot] = xroot
+            else:
+                parent[yroot] = xroot
+                rank[xroot] += 1
+
+        def kruskal(self2):
+
+            result = []
+            i, e = 0, 0
+            self2.graph = sorted(self2.graph, key=lambda item: item[2])
+            parent = []
+            rank = []
+            for node in range(self2.V):

```

Comment:
can you send the code where this is ported from? Seems there might be a more efficient way of doing things

Reply:
class Graph:
    def __init__(self, vertex):
        self.V = vertex
        self.graph = []
 
    def add_edge(self, u, v, w):
        self.graph.append([u, v, w])
 
 
    def search(self, parent, i):
        if parent[i] == i:
            return i
        return self.search(parent, parent[i])
 
    def apply_union(self, parent, rank, x, y):
        xroot = self.search(parent, x)
        yroot = self.search(parent, y)
        if rank[xroot] < rank[yroot]:
            parent[xroot] = yroot
        elif rank[xroot] > rank[yroot]:
            parent[yroot] = xroot
        else:
            parent[yroot] = xroot
            rank[xroot] += 1
 
  
    def kruskal(self):
        result = []
        i, e = 0, 0
        self.graph = sorted(self.graph, key=lambda item: item[2])
        parent = []
        rank = []
        for node in range(self.V):
            parent.append(node)
            rank.append(0)
        while e < self.V - 1:
            u, v, w = self.graph[i]
            i = i + 1
            x = self.search(parent, u)
            y = self.search(parent, v)
            if x != y:
                e = e + 1
                result.append([u, v, w])
                self.apply_union(parent, rank, x, y)
        for u, v, weight in result:
            print("Edge:",u, v,end =" ")
            print("-",weight)

Reply:
I meant a link. That is the same that is in your PR

Reply:
My bad, https://www.algotree.org/algorithms/minimum_spanning_tree/kruskals/

Reply:
Ok, make sure you heavily comment the code and move the graph class outside the test class

Reply:
Will do, finishing up some other edits before re-committing